### PR TITLE
[SPARK-24441][SS] Expose total estimated size of states in HDFSBackedStateStoreProvider

### PR DESCRIPTION
**Pull request by HeartSaVioR**
## What changes were proposed in this pull request?

This patch exposes the estimation of size of cache (loadedMaps) in HDFSBackedStateStoreProvider as a custom metric of StateStore.

The rationalize of the patch is that state backed by HDFSBackedStateStoreProvider will consume more memory than the number what we can get from query status due to caching multiple versions of states. The memory footprint to be much larger than query status reports in situations where the state store is getting a lot of updates: while shallow-copying map incurs additional small memory usages due to the size of map entities and references, but row objects will still be shared across the versions. If there're lots of updates between batches, less row objects will be shared and more row objects will exist in memory consuming much memory then what we expect. 

While HDFSBackedStateStore refers loadedMaps in HDFSBackedStateStoreProvider directly, there would be only one `StateStoreWriter` which refers a StateStoreProvider, so the value is not exposed as well as being aggregated multiple times. Current state metrics are safe to aggregate for the same reason.

## How was this patch tested?

Tested manually. Below is the snapshot of UI page which is reflected by the patch: 

screen shot 2018-06-05 at 10 16 16 pm

Please refer "estimated size of states cache in provider total" as well as "count of versions in state cache in provider".

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/HDFSBackedStateStoreProvider.scala [hunk 1]
```diff
@@ -181,6 +182,12 @@ private[state] class HDFSBackedStateStoreProvider extends StateStoreProvider wit
     }
   }
 
+  def getCustomMetricsForProvider(): Map[StateStoreCustomMetric, Long] = {
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/HDFSBackedStateStoreProvider.scala]: tiny nit:

```scala
def getCustomMetricsForProvider(): Map[StateStoreCustomMetric, Long] = synchronized {
  Map(metricProviderLoaderMapSize -> SizeEstimator.estimate(loadedMaps))
}
```

**Comment by HyukjinKwon** [comment]
Shall we make the PR title complete? Looks truncated.

**Comment by HeartSaVioR** [comment]
Thanks @HyukjinKwon for reviewing. Addressed PR title as well as fixing nit.

**Comment by jose-torres** [comment]
LGTM. To clarify the description, we expect the memory footprint to be much larger than query status reports in situations where the state store is getting a lot of updates?

**Comment by HeartSaVioR** [comment]
@jose-torres 
Ah yes I forgot that shallow copy has been occurring, so while new map should hold necessary size of map entries but row object will be shared across versions. Thanks for pointing it out. Will update the description.

**Comment by arunmahadevan** [comment]
@HeartSaVioR , may be then this should be reported in the "memoryUsedBytes" in the StateOperatorProgress (value reported in StreamingQueryProgress) because currently the usage reported does not reflect the memory used for the cache.

Question: in the screenshot "Estimated size of states cache in provider total" is 3.3 MB whereas the "memory used by state total" is 20.6 KB with "total number of state rows" = 2. This 150x difference is expected with just 2 rows in the state? Were there 100 versions of the map in the sample output you posted?

**Comment by HeartSaVioR** [comment]
@arunmahadevan 
I didn't add the metric to StateOperatorProgress cause this behavior is specific to HDFSBackedStateStoreProvider (though this is only one implementation available in Apache Spark) so not sure this metric can be treated as a general one. (@tdas what do you think about this?)

Btw, the cache is going to clean up when maintenance operation is in progress, so there could be more than 100 versions in map. Not sure why it shows 150x, but I couldn't find missing spot on the patch. Maybe the issue is from SizeEstimator.estimate()?

One thing we need to check is how SizeEstimator.estimate() calculate the memory usage when Unsafe row objects are shared across versions. If SizeEstimator adds the size of object whenever it is referenced, it will report much higher memory usage than actual.

**Comment by HeartSaVioR** [comment]
Also added custom metric for the count of versions stored in loadedMaps.

This is a new screenshot:
screen shot 2018-06-05 at 10 16 16 pm

**Comment by arunmahadevan** [comment]
>I didn't add the metric to StateOperatorProgress cause this behavior is specific to HDFSBackedStateStoreProvider

May be this can be reported as a custom metrics and keep it optional and that way its not tied to any specific implementation.

**Comment by HyukjinKwon** [comment]
@jose-torres is it good to go?

**Comment by HeartSaVioR** [comment]
@arunmahadevan 
Added custom metrics in state store to streaming query status as well. You can see `providerLoadedMapSize` is added to `stateOperators.customMetrics` in below output.

I have to exclude `providerLoadedMapCountOfVersions` from the list, since average metric is implemented a bit tricky and doesn't look like easy to aggregate for streaming query status. 
We may want to reimplement SQLMetric and subclasses to make sure everything works correctly without any tricky approach, but that doesn't look like trivial to address and I think this is out of scope on this PR.

```
18/06/06 22:51:23 INFO MicroBatchExecution: Streaming query made progress: {
  "id" : "7564a0b7-e3b2-4d53-b246-b774ab04e586",
  "runId" : "8dd34784-080c-4f86-afaf-ac089902252d",
  "name" : null,
  "timestamp" : "2018-06-06T13:51:15.467Z",
  "batchId" : 4,
  "numInputRows" : 547,
  "inputRowsPerSecond" : 67.15776550030694,
  "processedRowsPerSecond" : 65.94333936106088,
  "durationMs" : {
    "addBatch" : 7944,
    "getBatch" : 1,
    "getEndOffset" : 0,
    "queryPlanning" : 61,
    "setOffsetRange" : 5,
    "triggerExecution" : 8295,
    "walCommit" : 158
  },
  "eventTime" : {
    "avg" : "2018-06-06T13:51:10.313Z",
    "max" : "2018-06-06T13:51:14.250Z",
    "min" : "2018-06-06T13:51:07.098Z",
    "watermark" : "2018-06-06T13:50:36.676Z"
  },
  "stateOperators" : [ {
    "numRowsTotal" : 20,
    "numRowsUpdated" : 16,
    "memoryUsedBytes" : 26679,
    "customMetrics" : {
      "providerLoadedMapSize" : 181911
    }
  } ],
  "sources" : [ {
    "description" : "KafkaV2[Subscribe[apachelogs-v2]]",
    "startOffset" : {
      "apachelogs-v2" : {
        "2" : 489056,
        "4" : 489053,
        "1" : 489055,
        "3" : 489051,
        "0" : 489053
      }
    },
    "endOffset" : {
      "apachelogs-v2" : {
        "2" : 489056,
        "4" : 489053,
        "1" : 489055,
        "3" : 489051,
        "0" : 489053
      }
    },
    "numInputRows" : 547,
    "inputRowsPerSecond" : 67.15776550030694,
    "processedRowsPerSecond" : 65.94333936106088
  } ],
  "sink" : {
    "description" : "org.apache.spark.sql.execution.streaming.ConsoleSinkProvider@60999714"
  }
}
```

**Comment by arunmahadevan** [comment]
Nice, LGTM.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQueryListenerSuite.scala [hunk 2]
```diff
@@ -231,7 +231,7 @@ class StreamingQueryListenerSuite extends StreamTest with BeforeAndAfter {
   test("event ordering") {
     val listener = new EventCollector
     withListenerAdded(listener) {
-      for (i <- 1 to 100) {
+      for (i <- 1 to 50) {
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQueryListenerSuite.scala]: After the patch this test starts failing: it just means there's more time needed to run this loop 100 times, and doesn't mean the logic is broken. Decreasing number works for me.

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
ok to test

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQueryListenerSuite.scala (hunk 2 above)
Comment (jose-torres) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQueryListenerSuite.scala]: Makes sense, and I agree with the implicit claim that this slowdown isn't too worrying.

**Comment by jose-torres** [comment]
(Sorry to comment after so long with such a minor change - I've been busy with spark summit)

metricProviderLoaderMapSize should be metricProviderLoaderMapSizeBytes, both for clarity and consistency with memoryUsedBytes. Other than that LGTM.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/HDFSBackedStateStoreProvider.scala [hunk 3]
```diff
@@ -247,6 +253,14 @@ private[state] class HDFSBackedStateStoreProvider extends StateStoreProvider wit
   private lazy val fm = CheckpointFileManager.create(baseDir, hadoopConf)
   private lazy val sparkConf = Option(SparkEnv.get).map(_.conf).getOrElse(new SparkConf)
 
+  private lazy val metricProviderLoaderMapSizeBytes: StateStoreCustomSizeMetric =
+    StateStoreCustomSizeMetric("providerLoadedMapSizeBytes",
+      "estimated size of states cache in provider")
+
+  private lazy val metricProviderLoaderCountOfVersionsInMap: StateStoreCustomAverageMetric =
```
Comment (arunmahadevan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/HDFSBackedStateStoreProvider.scala]: Why is "metricProviderLoaderCountOfVersionsInMap" an average metrics? The other metrics like "numTotalStateRows" and even "providerLoadedMapSizeBytes" is count metric. Shouldn't this be similar?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala [hunk 4]
```diff
@@ -112,14 +122,19 @@ trait StateStoreWriter extends StatefulOperator { self: SparkPlan =>
     val storeMetrics = store.metrics
     longMetric("numTotalStateRows") += storeMetrics.numKeys
     longMetric("stateMemory") += storeMetrics.memoryUsedBytes
-    storeMetrics.customMetrics.foreach { case (metric, value) =>
-      longMetric(metric.name) += value
+    storeMetrics.customMetrics.foreach {
+      case (metric: StateStoreCustomAverageMetric, value) =>
+        longMetric(metric.name).set(value * 1.0d)
```
Comment (arunmahadevan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala]: How does this get accumulated ? It seems the value last set may get propagated to the driver.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/HDFSBackedStateStoreProvider.scala (hunk 3 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/HDFSBackedStateStoreProvider.scala]: This should be average to show `min, med, max` in SQL metrics UI, as I pasted capture of UI before. Summing them all doesn't give meaningful value.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala (hunk 4 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala]: If my understanding is right, the metric object (return of `longMetric()`) is different between each task, so the object will be different for each batch and each task. (TaskMetric is serialized and deserialized so it can't be shared between tasks.)

And actually the metric values are not aggregated into an SQLMetric object. The values are just aggregated and represented in SQLAppStatusListener.

https://github.com/apache/spark/blob/f5af86ea753c446df59a0a8c16c685224690d633/sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusListener.scala#L162-L174

https://github.com/apache/spark/blob/f5af86ea753c446df59a0a8c16c685224690d633/sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala#L147-L160

screen shot 2018-06-12 at 9 17 14 am

**Comment by HeartSaVioR** [comment]
Kindly ping again to @tdas

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala (hunk 4 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala]: We would be better to think about the actual benefit of exposing the value, rather than how to expose the value to somewhere. If we define it as count and do aggregation as summation, the aggregated value will be `(partition count * versions)` which might be hard for end users to find the meaning from the value.

I'm afraid that exposing this to StreamingQuery as average is not trivial, especially SQLMetric is defined as `AccumulatorV2[Long, Long]` so only single Long value can be passed. Under the restriction, we couldn't define `merge` operation for `average metric`.

**Comment by HeartSaVioR** [comment]
adding cc. to @zsxwing since he has been reviewing PRs for SS so far.

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by HeartSaVioR** [comment]
Rebased to fix conflict, and added new commit (last one: c9aada5) to represent cache hit / miss count in HDFS state provider. This is actually helpful for SPARK-24717 to determine proper value of configuration, but with this commit SPARK-24717 should be on top of this PR, so just added it here to avoid rebase hell.

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
Now I'm thinking about removing "metricProviderLoaderCountOfVersionsInMap" and also removing StateStoreCustomAverageMetric, since the value doesn't look correct with stream-stream join which handles multiple states in an operation.

We may try out SQLMetric to extend AccumulatorV2[(kind of Metric class), (kind of Metric class)] which Metric class can handle merging multiple values correctly according to the kind of metric, but this is beyond the scope of PR, so just removing "metricProviderLoaderCountOfVersionsInMap" would be clearer for this patch for now.

What do you think, @arunmahadevan @jose-torres ?

**Comment by jose-torres** [comment]
Sure, I don't mind if we remove that metric.

**Comment by HeartSaVioR** [comment]
retest this, please

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/progress.scala [hunk 5]
```diff
@@ -48,12 +49,24 @@ class StateOperatorProgress private[sql](
...
+      if (map.isEmpty) return JNothing
+      val keys = map.keySet.asScala.toSeq.sorted
+      keys.map { k => k -> valueToJValue(map.get(k)) : JObject }.reduce(_ ~ _)
+    }
+
+    val jsonVal = ("numRowsTotal" -> JInt(numRowsTotal)) ~
+      ("numRowsUpdated" -> JInt(numRowsUpdated)) ~
+      ("memoryUsedBytes" -> JInt(memoryUsedBytes))
+
+    if (!customMetrics.isEmpty) {
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/progress.scala]: You are already handling the case of map being empty in `safeMapToJValue` by adding JNothing. Doesnt JNothing values just get dropped from the json text any way?

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/progress.scala [hunk 6]
```diff
@@ -48,12 +49,24 @@ class StateOperatorProgress private[sql](
...
 
   private[sql] def copy(newNumRowsUpdated: Long): StateOperatorProgress =
-    new StateOperatorProgress(numRowsTotal, newNumRowsUpdated, memoryUsedBytes)
+    new StateOperatorProgress(numRowsTotal, newNumRowsUpdated, memoryUsedBytes, customMetrics)
 
   private[sql] def jsonValue: JValue = {
-    ("numRowsTotal" -> JInt(numRowsTotal)) ~
-    ("numRowsUpdated" -> JInt(numRowsUpdated)) ~
-    ("memoryUsedBytes" -> JInt(memoryUsedBytes))
+    def safeMapToJValue[T](map: ju.Map[String, T], valueToJValue: T => JValue): JValue = {
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/progress.scala]: T is always Long. Why make a generic function for that? This does not even need a separate function.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala [hunk 7]
```diff
@@ -81,10 +81,10 @@ class SQLMetric(val metricType: String, initValue: Long = 0L) extends Accumulato
...
 
 object SQLMetrics {
-  private val SUM_METRIC = "sum"
-  private val SIZE_METRIC = "size"
-  private val TIMING_METRIC = "timing"
-  private val AVERAGE_METRIC = "average"
+  val SUM_METRIC = "sum"
+  val SIZE_METRIC = "size"
+  val TIMING_METRIC = "timing"
+  val AVERAGE_METRIC = "average"
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala]: why this change?

**Review by tdas — CHANGES_REQUESTED** [review]
This looks good. Only two comments.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala (hunk 7 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala]: It was to handle exception case while aggregating custom metrics, especially filtering out average since it is not aggregated correctly. Since we remove custom average metric, we no longer need to filter out them. Will revert the change as well as relevant logic.

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/progress.scala (hunk 6 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/streaming/progress.scala]: I've first trying to leverage `StreamingQueryProgress.safeMapToJValue` but can't find proper place to move to be co-used, so I simply copied it. Will simplify the code block and inline.

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/progress.scala (hunk 5 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/streaming/progress.scala]: Actually didn't notice that. Thanks for letting me know! Will simplify.

**Comment by HeartSaVioR** [comment]
@tdas Thanks for the review! Addressed review comments.

**Comment by HeartSaVioR** [comment]
Retest this, please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HeartSaVioR** [comment]
@tdas Kindly reminder.

**Comment by tdas** [comment]
I am having a second thoughts about this. Exposing the entire memory usage of all the loaded maps as another custom metric .... just adds more confusion. Rather the point of the the main state metric `memoryUsedBytes` is to capture how much memory is occupied because of the one partition of the state, and that implicitly should cover all the loaded versions of that state partition. So I strongly feel that instead of adding a custom metric, we should change the existing `memoryUsedBytes` to capture all the memory. 

I am fine adding the custom metrics hit and miss counts. No questions about that. 

What do you think?

**Comment by HeartSaVioR** [comment]
My series of patches could be possible based on two metrics: `size for memory usage of latest version` and `size for total memory usage of loaded versions`. SPARK-24717 (#21700) enabled the possibility to tune the overall state memory usage, and if end users have either one metric they couldn't tune it.

IMHO, I'm not 100% sure how much this patch provides confusion to the end users, but if the intention of `memoryUsedBytes` is for measuring overall state partition, what about replacing `memoryUsedBytes` as `size for total memory usage of loaded versions`, but also placing `size for memory usage of latest version` to custom metric?

**Comment by tdas** [comment]
@HeartSaVioR I think I agree with a second approach that you suggested. So 
`memoryUsedBytes` => `size for total memory usage of loaded versions` and
`customMetric` => `size for memory usage of latest version`

**Comment by HeartSaVioR** [comment]
@tdas Thanks for the feedback! Updated the PR.

**Comment by tdas** [comment]
LGTM.

**Comment by tdas** [comment]
Merged to master.

**Comment by HeartSaVioR** [comment]
Thanks all for reviewing and thanks @tdas for merging this in!

**Comment by tdas** [comment]
Unfortunately this PR broke the master build. Looks like some import that probably got removed in the other PR I merged, which didnt create any direct conflict.

**Comment by HeartSaVioR** [comment]
@tdas In case of you are not working on the patch, I'm working on the fix and will provide minor PR.

**Comment by tdas** [comment]
I did. Fixed the import
